### DISCUSSION TITLE
FastAPI response_model unable to handle Iterable types -- returns only the model keys

### DISCUSSION THREAD

**Discussion by interifter** [fastapi/fastapi#8829] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Any, Iterable, List
from fastapi import FastAPI, APIRouter
from pydantic import BaseModel

# Define simple model
class Simple(BaseModel):
    spam: int
    eggs: str

# instantiate fast api and route
loader = FastAPI(
    version="0",
)

requests_router = APIRouter(prefix="/requests", tags=["requests"])

@requests_router.get("/", response_model=Iterable[Simple])
def get_requests() -> Iterable[Simple]:
    return [
        Simple(
            spam=1,
            eggs="a"
        )
    ]

loader.include_router(requests_router)

###
### Output at http://127.0.0.1:8000/requests/ is 
### {
###     "spam": "eggs"
### }
###
```

### Description

To reproduce

* Run uvicorn api2:loader --reload
* Go to http://127.0.0.1:8000/requests/
* Observe only the keys in the data model are returned

To fix
* Change the `response_model` to `response_model=List[Simple]`
* Go to http://127.0.0.1:8000/requests/
* Observe you get the correct response:
```json
[
    {
        "spam": 1,
        "eggs": "a"
    }
]
```

I haven't been able to find this in any documentation or bugs, but maybe my google skills are lacking

```
uvicorn            0.15.0
fastapi            0.68.1
```

### Operating System

Windows

### Operating System Details

Windows 11 21H2 (OS Build 22000.160))

### FastAPI Version

0.68.1

### Python Version

3.9.1

### Additional Context

This has a workaround, but I don't know the code base well enough to understand why an `Iterable` type would cause this problem

**Comment by interifter** [comment, marked as answer]
This was the original problem I thought I had: #679

However, this issue occurs when I use `Iterable` in my pydantic data model:

```python
class PrecheckRequest(BaseModel):
    job_id: int = Field(alias="jobId")
    status: StatusType #Enum type
    callback: Optional[Iterable[str]] = Field(default=[]) # This causes problems

@router.get("/", response_model=List[PrecheckRequest])
def get_requests() -> Iterable[PrecheckRequest]:
    # Do work

```

Results in:
```
ValueError: [ValueError('dictionary update sequence element #0 has length 14; 2 is required'), TypeError('vars() argument must have __dict__ attribute')]
...
in fastapi\encoders.py", line 144
```



When I change `Iterable` to `List` in the pydantic data model, it starts working as desired

```python
class Request(BaseModel):
    job_id: int = Field(alias="jobId")
    status: StatusType #Enum type
    callback: Optional[List[str]] = Field(default=[]) # This fixes the problem

@router.get("/", response_model=List[Request])
def get_requests() -> Iterable[Request]:
    # Do work

```

**Comment by interifter** [comment]
Another bug I thought was potentially the problem: #2813

**Comment by klaa97** [comment]
The main issue is that the `jsonable_encoder` doesn't support raw `Iterator` (produced by Pydantic when validating using as a response_model `Iterable[Simple]`), but it does support List .

I opened #3913 to try to fix the behavior - I tested in local your example code and it should work as you expect.

**Comment by interifter** [comment]
@klaa97 it looks like a very straightforward fix, from the PR.

**Comment by rojvv** [comment]
Need this

**Comment by interifter** [comment]
> Need this

It looks like we just got the PR for it merged. #3913
